File: src/radarc/dem.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Iterator, Tuple

import numpy as np


@dataclass(frozen=True)
class RadarSite:
    """Location of a radar expressed in a local cartesian grid."""

    x_m: float
    y_m: float
    height_agl_m: float


class DigitalElevationModel:
    """Bilinearly interpolated DEM backed by a numpy grid."""

    def __init__(
        self,
        elevations: np.ndarray,
        *,
        origin_x_m: float,
        origin_y_m: float,
        cell_size_m: float,
        no_data_value: float | None = None,
        description: str | None = None,
    ) -> None:
        if elevations.ndim != 2:
            raise ValueError("elevations must be a 2-D array")
        self._grid = elevations.astype(float)
        self.origin_x_m = float(origin_x_m)
        self.origin_y_m = float(origin_y_m)
        self.cell_size_m = float(cell_size_m)
        self.no_data_value = no_data_value
        self.description = description or ""

    @property
    def shape(self) -> Tuple[int, int]:
        return self._grid.shape

    @property
    def width(self) -> int:
        return self._grid.shape[1]

    @property
    def height(self) -> int:
        return self._grid.shape[0]

    @property
    def extent(self) -> Tuple[float, float, float, float]:
        """Return (min_x, min_y, max_x, max_y)."""
        max_x = self.origin_x_m + (self.width - 1) * self.cell_size_m
        max_y = self.origin_y_m + (self.height - 1) * self.cell_size_m
        return (self.origin_x_m, self.origin_y_m, max_x, max_y)

    @classmethod
    def from_npz(cls, path: str | Path) -> "DigitalElevationModel":
        data = np.load(path)
        grid = data["elevations"]
        origin_x_m = float(data["origin_x_m"])
        origin_y_m = float(data["origin_y_m"])
        cell_size_m = float(data["cell_size_m"])
        description = str(data["description"]) if "description" in data else None
        return cls(
            grid,
            origin_x_m=origin_x_m,
            origin_y_m=origin_y_m,
            cell_size_m=cell_size_m,
            no_data_value=float(data["no_data_value"]) if "no_data_value" in data else None,
            description=description,
        )

    def contains(self, x_m: float, y_m: float) -> bool:
        min_x, min_y, max_x, max_y = self.extent
        return min_x <= x_m <= max_x and min_y <= y_m <= max_y

    def sample(self, x_m: float, y_m: float) -> float | None:
        """Return bilinearly interpolated elevation (meters)."""
        fx = (x_m - self.origin_x_m) / self.cell_size_m
        fy = (y_m - self.origin_y_m) / self.cell_size_m
        if fx < 0 or fy < 0 or fx > self.width - 1 or fy > self.height - 1:
            return None

        x0 = int(np.floor(fx))
        y0 = int(np.floor(fy))
        x1 = min(x0 + 1, self.width - 1)
        y1 = min(y0 + 1, self.height - 1)
        dx = fx - x0
        dy = fy - y0

        q11 = self._grid[y0, x0]
        q21 = self._grid[y0, x1]
        q12 = self._grid[y1, x0]
        q22 = self._grid[y1, x1]

        if self._is_nodata(q11, q21, q12, q22):
            return None

        r1 = (1 - dx) * q11 + dx * q21
        r2 = (1 - dx) * q12 + dx * q22
        return (1 - dy) * r1 + dy * r2

    def _is_nodata(self, *values: float) -> bool:
        if self.no_data_value is None:
            return False
        return any(np.isclose(v, self.no_data_value) for v in values)

    def grid_points(self) -> Iterator[Tuple[float, float, float]]:
        for row in range(self.height):
            y = self.origin_y_m + row * self.cell_size_m
            for col in range(self.width):
                x = self.origin_x_m + col * self.cell_size_m
                yield x, y, self._grid[row, col]

    def profile_between(
        self,
        start: Tuple[float, float],
        end: Tuple[float, float],
        step_m: float | None = None,
    ) -> Iterable[Tuple[float, float]]:
        """Yield (distance_along_m, terrain_height_m)."""
        sx, sy = start
        ex, ey = end
        dx = ex - sx
        dy = ey - sy
        total_distance = float(np.hypot(dx, dy))
        if total_distance == 0:
            elevation = self.sample(sx, sy)
            if elevation is None:
                return []
            return [(0.0, elevation)]

        if step_m is None:
            step_m = self.cell_size_m / 2
        steps = max(int(np.ceil(total_distance / step_m)), 2)
        samples: list[Tuple[float, float]] = []
        for i in range(steps + 1):
            t = i / steps
            x = sx + t * dx
            y = sy + t * dy
            elevation = self.sample(x, y)
            if elevation is None:
                continue
            samples.append((t * total_distance, elevation))
        return samples

File: src/dem.rs
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::{Path, PathBuf};

use ndarray::Array2;
use serde::Deserialize;
use thiserror::Error;

#[derive(Debug, Clone, Copy)]
pub struct RadarSite {
    pub x_m: f64,
    pub y_m: f64,
    pub height_agl_m: f64,
}

#[derive(Debug, Error)]
pub enum DemError {
    #[error("failed to read DEM file: {0}")]
    Io(#[from] std::io::Error),
    #[error("failed to parse DEM json: {0}")]
    Parse(#[from] serde_json::Error),
    #[error("invalid HGT file size: {0}")]
    InvalidHgtSize(usize),
    #[error("failed to parse HGT filename for geo origin: {0}")]
    HgtName(PathBuf),
    #[error("DEM grid must be rectangular")]
    NonRectangular,
    #[error("DEM grid must be at least 1x1")]
    Empty,
}

#[derive(Debug, Deserialize)]
struct DemFile {
    origin_x_m: f64,
    origin_y_m: f64,
    cell_size_m: f64,
    elevations: Vec<Vec<f64>>,
    #[serde(default)]
    no_data_value: Option<f64>,
    #[serde(default)]
    description: Option<String>,
}

#[derive(Debug, Clone)]
pub struct DigitalElevationModel {
    grid: Array2<f64>,
    pub origin_x_m: f64,
    pub origin_y_m: f64,
    pub cell_size_m: f64,
    pub no_data_value: Option<f64>,
    pub description: Option<String>,
}

impl DigitalElevationModel {
    pub fn from_json_file<P: AsRef<Path>>(path: P) -> Result<Self, DemError> {
        let file = File::open(path)?;
        let reader = BufReader::new(file);
        let data: DemFile = serde_json::from_reader(reader)?;
        Self::from_dem_file(data)
    }

    pub fn from_hgt_file<P: AsRef<Path>>(path: P) -> Result<Self, DemError> {
        let path_ref = path.as_ref();
        let mut file = File::open(path_ref)?;
        let mut buf = Vec::new();
        file.read_to_end(&mut buf)?;
        let bytes = buf.len();
        let side = match bytes / 2 {
            n if n == 1201 * 1201 => 1201usize,
            n if n == 3601 * 3601 => 3601usize,
            _ => return Err(DemError::InvalidHgtSize(bytes)),
        };

        use byteorder::{BigEndian, ReadBytesExt};
        let mut cursor = std::io::Cursor::new(&buf);
        let mut values: Vec<f64> = Vec::with_capacity(side * side);
        for _ in 0..(side * side) {
            let v = cursor.read_i16::<BigEndian>()?;
            if v == -32768 {
                values.push(f64::NAN);
            } else {
                values.push(v as f64);
            }
        }
        let grid = Array2::from_shape_vec((side, side), values).expect("shape ok");

        // Parse SW corner from filename like N37W122.hgt
        let name = path_ref
            .file_stem()
            .and_then(|s| s.to_str())
            .ok_or_else(|| DemError::HgtName(path_ref.to_path_buf()))?;
        let (lat_sign, rest) = name.split_at(1);
        let (lat_deg_s, rest) = rest.split_at(2);
        let (lon_sign, lon_deg_s) = rest.split_at(1);
        let lat_deg: f64 = lat_deg_s.parse().map_err(|_| DemError::HgtName(path_ref.to_path_buf()))?;
        let lon_deg: f64 = lon_deg_s
            .chars()
            .take_while(|c| c.is_ascii_digit())
            .collect::<String>()
            .parse()
            .map_err(|_| DemError::HgtName(path_ref.to_path_buf()))?;
        let lat0 = if lat_sign.eq_ignore_ascii_case("S") { -lat_deg } else { lat_deg };
        let lon0 = if lon_sign.eq_ignore_ascii_case("W") { -lon_deg } else { lon_deg };

        // Approx meters per degree at tile latitude
        let meters_per_deg_lat = 111_320.0f64;
        let meters_per_deg_lon = meters_per_deg_lat * (lat0.to_radians().cos());
        let samples_per_degree = (side - 1) as f64;
        let step_lon_deg = 1.0 / samples_per_degree;
        let step_lat_deg = 1.0 / samples_per_degree;
        // Use average step size as square cell size approximation
        let cell_size_m = ((step_lat_deg * meters_per_deg_lat) + (step_lon_deg * meters_per_deg_lon)) * 0.5;

        Ok(Self {
            grid,
            origin_x_m: 0.0, // treat SW corner as (0,0) in meters
            origin_y_m: 0.0,
            cell_size_m,
            no_data_value: Some(f64::NAN),
            description: Some(format!("HGT tile at lat {}, lon {}", lat0, lon0)),
        })
    }

    pub fn from_rows(
        elevations: Vec<Vec<f64>>,
        origin_x_m: f64,
        origin_y_m: f64,
        cell_size_m: f64,
    ) -> Result<Self, DemError> {
        let file = DemFile {
            origin_x_m,
            origin_y_m,
            cell_size_m,
            elevations,
            no_data_value: None,
            description: None,
        };
        Self::from_dem_file(file)
    }

    fn from_dem_file(file: DemFile) -> Result<Self, DemError> {
        let height = file.elevations.len();
        if height == 0 {
            return Err(DemError::Empty);
        }
        let width = file.elevations[0].len();
        if width == 0 {
            return Err(DemError::Empty);
        }
        for row in &file.elevations {
            if row.len() != width {
                return Err(DemError::NonRectangular);
            }
        }
        let flattened: Vec<f64> = file.elevations.into_iter().flatten().collect();
        let grid = Array2::from_shape_vec((height, width), flattened).expect("validated dims");
        Ok(Self {
            grid,
            origin_x_m: file.origin_x_m,
            origin_y_m: file.origin_y_m,
            cell_size_m: file.cell_size_m,
            no_data_value: file.no_data_value,
            description: file.description,
        })
    }

    pub fn width(&self) -> usize {
        self.grid.ncols()
    }

    pub fn height(&self) -> usize {
        self.grid.nrows()
    }

    pub fn extent(&self) -> (f64, f64, f64, f64) {
        let max_x = self.origin_x_m + (self.width() as f64 - 1.0) * self.cell_size_m;
        let max_y = self.origin_y_m + (self.height() as f64 - 1.0) * self.cell_size_m;
        (self.origin_x_m, self.origin_y_m, max_x, max_y)
    }

    pub fn contains(&self, x_m: f64, y_m: f64) -> bool {
        let (min_x, min_y, max_x, max_y) = self.extent();
        x_m >= min_x && x_m <= max_x && y_m >= min_y && y_m <= max_y
    }

    pub fn sample(&self, x_m: f64, y_m: f64) -> Option<f64> {
        let fx = (x_m - self.origin_x_m) / self.cell_size_m;
        let fy = (y_m - self.origin_y_m) / self.cell_size_m;

        if fx < 0.0 || fy < 0.0 {
            return None;
        }
        let max_x = self.width() as f64 - 1.0;
        let max_y = self.height() as f64 - 1.0;
        if fx > max_x || fy > max_y {
            return None;
        }

        let x0 = fx.floor() as usize;
        let y0 = fy.floor() as usize;
        let x1 = (x0 + 1).min(self.width() - 1);
        let y1 = (y0 + 1).min(self.height() - 1);
        let dx = fx - x0 as f64;
        let dy = fy - y0 as f64;

        let q11 = self.grid[(y0, x0)];
        let q21 = self.grid[(y0, x1)];
        let q12 = self.grid[(y1, x0)];
        let q22 = self.grid[(y1, x1)];

        if self.is_nodata(q11) || self.is_nodata(q21) || self.is_nodata(q12) || self.is_nodata(q22)
        {
            return None;
        }

        let r1 = (1.0 - dx) * q11 + dx * q21;
        let r2 = (1.0 - dx) * q12 + dx * q22;
        Some((1.0 - dy) * r1 + dy * r2)
    }

    pub fn grid_points(&self) -> GridPointIter<'_> {
        GridPointIter {
            dem: self,
            row: 0,
            col: 0,
        }
    }

    pub fn elevation_value(&self, row: usize, col: usize) -> f64 {
        self.grid[(row, col)]
    }

    pub fn profile_between(
        &self,
        start: (f64, f64),
        end: (f64, f64),
        step_m: Option<f64>,
    ) -> Vec<(f64, f64)> {
        let (sx, sy) = start;
        let (ex, ey) = end;
        let dx = ex - sx;
        let dy = ey - sy;
        let total_distance = (dx * dx + dy * dy).sqrt();
        if total_distance == 0.0 {
            return self
                .sample(sx, sy)
                .map(|e| vec![(0.0, e)])
                .unwrap_or_default();
        }
        let step = step_m.unwrap_or(self.cell_size_m / 2.0);
        let steps = ((total_distance / step).ceil() as usize).max(2);
        let mut samples = Vec::with_capacity(steps + 1);
        for i in 0..=steps {
            let t = i as f64 / steps as f64;
            let x = sx + t * dx;
            let y = sy + t * dy;
            if let Some(elev) = self.sample(x, y) {
                samples.push((t * total_distance, elev));
            }
        }
        samples
    }

    fn is_nodata(&self, value: f64) -> bool {
        self.no_data_value
            .map(|v| (value - v).abs() < f64::EPSILON)
            .unwrap_or(false)
    }
}

pub struct GridPointIter<'a> {
    dem: &'a DigitalElevationModel,
    row: usize,
    col: usize,
}

impl<'a> Iterator for GridPointIter<'a> {
    type Item = (f64, f64, f64);

    fn next(&mut self) -> Option<Self::Item> {
        if self.row >= self.dem.height() {
            return None;
        }
        let x = self.dem.origin_x_m + self.col as f64 * self.dem.cell_size_m;
        let y = self.dem.origin_y_m + self.row as f64 * self.dem.cell_size_m;
        let value = self.dem.grid[(self.row, self.col)];

        self.col += 1;
        if self.col >= self.dem.width() {
            self.col = 0;
            self.row += 1;
        }

        Some((x, y, value))
    }
}
